### ISSUE TITLE
Change axes_manager indexing feature

### ISSUE THREAD

**Issue by francisco-dlp** [hyperspy/hyperspy#814]
Currently it is possible to index the `AxesManager` using complex number. From the docstring:

```In addition AxesManager support indexing using
complex numbers a + bj, where b can be one of 0, 1, 2 and 3 and a a valid
index. If b is 3 AxesManager is indexed using the order of the axes in the
array. If b is 1(2), indexes only the navigation(signal) axes in the
natural order. ```

However, #388 proposes to avoid mixing signal and navigation indexing as this can be confusing. I propose to extend this to the `AxesManager` indexing feature by replacing the previous convention with a simpler (and more limited) syntax: real numbers to index the navigation axes, pure imaginary numbers for signal indexing. What do you think?

**Comment by vidartf** [comment]
Complicated issue. I agree this could benefit from a makeover, as there is no intuitive logic behind having 0j be natural order, 1j be nav, 2j signal and 3j be array order. This makes it difficult to remember, and makes existing code harder to read. On the other hand, accessing axis X in a specific order is really useful. It is also useful to have a single function that can take an index in any kind of order, as well as an axis name (and I'd which that it would check if the argument is a `DataAxis` as well, which could simply be returned as it is).

So for me, I'm all for looking for a better solution, but I think we should brain-storm a little to see if we can't find a solution that gets rid of the cons, but keeps all/most the pros of the existing way.

**Comment by to266** [comment]
1. I agree that using real(imaginary) indices to index navigation(signal) axes is going in the right direction
2. Nevertheless, for internal usage I quite like that it's possible to to index the axes in array order using public methods, even though I do not use it because I always forget it exists..

...as a side-note/joke, prof. Gull, while teaching us dynamics, mentioned [this](http://geometry.mrao.cam.ac.uk/wp-content/uploads/2015/02/ImagNumbersArentReal.pdf). Shall we just make use of it and have all three nicely separated? ;)

**Comment by to266** [comment]
(as for the Geometric algebra, I may be misremembering things.... so it might not be as relevant as I hoped. An interesting read nevertheless)

**Comment by francisco-dlp** [comment]
The raison d'etre of that feature was provide a way to express e.g. "the first signal axis" in the method that take an axis argument e.g. `s.sum(axis=0+2j)`. This  is retained in the proposed syntax. I think that we should avoid to mix in signal and navigation indexing in HyperSpy's code, except in the cases where it is fully unavoidable or much more convenient. Where this is the case, then we can simply access `AxesManager._axes`. Could you provide an example where the syntax that this proposes to remove is necessary or much more convenient?

**Comment by vidartf** [comment]
Well, `AxesManager._axes` is made protected for a reason: It's a list we don't want people to interact with directly, as its order and content is important for the correct operation of `AxesManager`. A possible solution to this is to expose a read-only property (i.e. only getter) `AxesManager.axes` that simply returns a tuple variant of `AxesManager._axes`.

Given this, we could of course also use `AxesManager.signal_axes`, `AxesManager.navigation_axes` and `AxesManager.nat_axes` (or similar). These are definitely easier to read, but are maybe less concise and elegant.

From my personal use, I spend much more time mucking about trying to keep the bookkeeping of natural order vs. array order correct (and `DataAxis.index_in_array` etc is only reliable if the `DataAxis` is not shared between several `AxesManager`). So I'm not sure it makes sense to cut those two at the expense of signal/navigation.

**Comment by francisco-dlp** [comment]
I agree that we should definitely encourage using `AxesManager.signal_axes` and `AxesManager.navigation_axes`. I also think that it an excellent good idea to create an `axes` property (maybe `data_axes` to stress the fact that they are in the array order?). I am not so sure about adding `AxesManager.nat_axes` because the point of #388 was to avoid mixing signal and navigation indexing where it can create confusion. If someone needs it he/she can simply index   `AxesManager.navigation_axes + `AxesManager.signal_axes` which is more explicit.

In any case, the main point of this issue is independent of those enhancements: indexing the signal and navigation axes independently using one single `axis` argument. I still think that real for navigation and pure complex for signal is better (in the sense of #388) than the current syntax.

**Comment by vidartf** [comment]
Looking over my code where I actually use `3j` indexing, I feel most of my reluctance can be resolved by having `AxesManager.__getitem__` accept `DataAxis` arguments. If the `DataAxis` is in the `AxesManager`, simply return it, if not, raise a `ValueError`.

The pattern I'm trying to avoid is like this:
```python
for ax in axes_collection:
    signal = signal.sum(ax.index_in_array + 3j)
```
This would be a lot better, and would be necessary with the proposed changes:
```python
for ax in axes_collection:
    signal = signal.sum(ax)
```

Other than that, if a PR implementing this is able to reform all the existing code to use the new API in a nice way, I will not have any objections. If the PR has a lot of repeated code to solve specific tasks, that code needs to be made into a function somehow.

If we are discussing which space should be real, and which imaginary, then neither makes great intuitive sense. If we're reducing the ways to directly index `AxesManager`, then it should only be natural order: When indexing/slicing `AxesManager`, you expect to index/slice *all* axes it has. If you only want to slice navigation axes, then `AxesManager.navigation_axes` is the way to go. Having any other usage pattern is *at least* as confusing as mixing navigation/signal axes in one indexing operation, likely more (there is nothing intuitive about using imaginary numbers for indexing, and I'm reasonably sure it prevents slicing).

**Comment by francisco-dlp** [comment]
Adding `DataAxis` to the valid arguments makes a lot of sense.

I disagree on indexing only using the natural order. As I mentioned before, the syntax must be able to express something like "the first signal axis". There is no way to express this using the "mixed natural order" syntax. HyperSpy is all about into classifying axes in two sets, navigation and signal. Finding an intuitive syntax to operate using these sets  is important, and I think that I got it wrong with the `AxesManager` syntax because it is unnecessarily rich and not easy to remember (I have to look at the docstring every time that I want to use it!). The one that I propose now is a simplification of the existing one. It still has the drawback that using real for nav and imag for sig is not terribly intuitive, but at least it is simpler and easier to remember than the current one. We have always placed the nav axes first and in most cases people tend to put the real part of a complex number first.

**Comment by vidartf** [comment]
If we're keeping complex number indexing, then yes, nav=real and sig=imaginary makes the most sense. However, the syntax does not have to be able to express "first signal axis", as the more verbose and straight-forward way to do it (if we accept `DataAxis`) would be to do `signal.sum(axis=s.axes_manager.signal_axes[0])`. This would also be consistent with the forced use of `inav`/`isig` (longer and more verbose, but clear). I'm not saying I'm necessarily all for this, but the same logic as for `isig`/`inav` applies here, and we've already made that decision.

**Comment by francisco-dlp** [comment]
You are right, `signal.sum(axis=signal.axes_manager.signal_axes[0])` would be equivalent and clearer. Then the question is: should we simplify the syntax as proposed above or disable integer indexing all together? (I think that we agree in that keeping the natural mixed syntax is evil.) My personal preference is to provide the more compact compact syntax too as I find the other one too verbose, but I realise that this is against the spirit of #388. What do others think?

**Comment by vidartf** [comment]
I agree that keeping some sort of short-hand is really, really useful, but I'm not convinced that imaginary numbers is the way to go. How about tuple indexing:
```python
signal.axes_manager['sig', 0]
signal.sum(axis=('sig', 0))
```
Then we could also keep `'nat'` and `'data'` around at little/no extra cost. We could possibly also work with reserved axes names, as in #422:
```python
signal.sum('signal0')
signal.sum('s0')
# or something similar
```
If we go for tuples, then we can resolve the issue in #422 by specifying all navigation axes as `('nav', None)` (alternatively `('nav', slice(None))`), i.e. without reserving axis name `'navigation'`/`'signal'`.

PS: I was wrong about not being able to use complex numbers for slicing.

**Comment by to266** [comment]
Do we do anything here after all?

**Comment by francisco-dlp** [comment]
I think that memorizing that real is for navigation and imag for signal is easier than memorising the tuple syntax that @vidartf proposes. I don't think that keeping the ability to index in the "natural" and "array" order is an advantage of this approach, as we're aiming at not exposing "mixed" indexing to the users.

**Comment by vidartf** [comment]
I'm for dropping number indexing at all (or just keeping it as it is for backwards compatibility), and just supporting name or `DataAxis`, in the spirit of getting closer to "one obvious way of completing the task". Either way I'm for adding `data_axes` property. Having natural order property would also be nice, as it is way too long to write `signal.axes_manager.navigation_axes + signal.axes_manager.signal_axes` (it basically fills a complete code line on its own).

Either way I think the tuple suggestion of mine was superseded by #823.

**Comment by francisco-dlp** [comment]
We cannot drop number indexing because we need to be able to express things like the last navigation axis without using properties of a given `AxesManager`. `s.method(axis=s.axes_manager.navigation_axes[-1])` is not only too verbose for what is a common task, also it cannot be used to set the default value of the argument. We could default to `None` and then define what `None` means, but that is less explicit than the complex syntax.

**Comment by vidartf** [comment]
Then as before: As long as a PR can be made to implement this and replace all the existing use in an elegant matter, I'll be happy. :)
